Introduce structured, span-based observability through Logger interface

wvanlint opened the pull request:
This change allows users to create hierarchical span objects through the `Logger` interface for specific computations, such as the handling of HTLCs. These span objects will be held in LDK across the corresponding lifetimes before being dropped, providing insight in durations and latencies.

This API is designed to be compatible with https://docs.rs/opentelemetry/latest/opentelemetry/trace/trait.Tracer.html, but can also be used more directly to derive time-based metrics or to log durations.

Hierarchical RAII spans are currently added for:
- HTLC lifetimes, HTLC state transitions, inbound to outbound HTLC forwarding (see `functional_tests.rs`).
- Ping/pong request-response pairs.

wvanlint committed:
Introduce structured, span-based observability through Logger interface

This change allows users to create hierarchical span objects through the
Logger interface for specific computations, such as the handling of
HTLCs. These span objects will be held in LDK across the corresponding
lifetimes before being dropped, providing insight in durations and
latencies.

TheBlueMatt reviewed (commented):
Oops, okay so there was def a miscommunication when we spoke about switching to an RAII span concept :sweat_smile:. At a high level I wasn't anticipate we'd give spans for a total HTLC forwards - I was assuming we'd only do spans for basically function runtime in LDK - "how long did it take us to process events", "how long did it take us to process the pending forwards", etc. From there, we might give spans for things like monitor updates (including the parent idea, which is cool), so you could answer questions like "how much of the time spent processing pending forwards is spent in monitor updates".

By focusing on answering questions about specific function runtime the "parent" thing is super well-defined. But once we switch to non-functional spans, it becomes a bit murky - a monitor update might block several payment forwards, so should it have all of them as "parents"?

Of course using only functional blocks as spans makes it less able to answer the types of questions I think you specifically want, eg "how long is it taking us to forward most HTLCs", but as it lies its hard to take that and answer "why is it taking us long to forward some HTLCs".

We could also split the difference and have functional spans for LDK things, but also non-functional spans for outside-LDK things - ie the above spans plus a span for how long each async monitor update took to persist as well as how long it took to get a response from a peer once we send a message that needs a response. Sadly the latter is a bit annoying to build, commitment_signed -> revoke_and_ack latency is pretty easy to measure, though once we get the RAA we're still waiting on another commitment_signed in many cases. There may be a way to do it specific to an HTLC kinda like you have here but just around messages themselves. 

WDYT?

wvanlint commented:
Haha no worries, I imagined it it wasn't the same but wanted to write this up as a starting point. 

I agree on the difference between function scope spans and higher-level state machine spans and how they answer different questions. Given the amount of logic that depends on messages, timers, and asynchronous persistence completions, I think we would need both complementing each other to have a complete understanding.
- Function scope spans are clear and easy to optimize for, but don't cross certain boundaries and would not give a complete picture. User trait implementations are already easy to measure but it's difficult to decide when it's impactful to improve them.
- Higher-level spans around the Lightning state machine would give an end-to-end view for Lightning payments and can identify rough areas of investigation. 

If we're building higher-level state machine spans on top of messages outside of LDK as in https://discord.com/channels/915026692102316113/978829624635195422/1296877062015029258, would it lead to rebuilding/duplicating something similar to the Lightning state machine to get the full picture? It seems like `channel.rs` already has all the necessary information.

I'm also curious what we should optimize for. I'm assuming we would like to have maximal observability, while minimizing overhead. Is maintaining span objects in the Lightning state machine too invasive? Would there be a more minimal event API that can be added to `channel.rs`?

Having `start(span: Span, parent: Option<&Span>) -> ()` and `end(span: Span) -> ()` methods without the RAII concept could be less invasive, possibly with some helpers. We could also still verify in `TestLogger` whether all spans are ended correctly during tests.

wvanlint committed:
Add WaitingOn* spans to inbound HTLCs and forwards

wvanlint committed:
Add WaitingOn* spans to outbound HTLCs

wvanlint commented:
Apologies for the delay here, a project at work came up that needed to be prioritized. 

I rebased the PR and added additional `WaitingOn*` spans as leafs in the HTLC span hierarchy. I still need to verify completeness there.

> It's pretty gnarly to actually answer the why, here, sadly, though - you have to track the HTLC:channel relation and then track time in the channel "states" during the time the HTLC is in different states...I suppose we could break out the HTLC state transitions into channel sub-states

Right, the HTLC:channel relation can't be represented simply, but you can indeed break out the spans from the viewpoint of an HTLC, which feels like the right approach to me. If an HTLC needs to wait for a remote revocation, an HTLC-specific span can be started at that point in time until it arrives, even though other HTLCs might have started waiting for that revocation beforehand. These spans would be specifically from the viewpoint of each HTLC, but I believe it does explain the why? When complete, the `WaitingOn*` leaf spans should fully cover the lifecycle of an HTLC.

The spans from the viewpoint of each HTLC might be complemented by other metrics/spans, such as measuring the duration of `Persist` calls more directly.

wvanlint commented on lightning/src/util/logger.rs:
```diff
@@ -160,8 +161,105 @@ impl_record!(, 'a);
 	fn log(&self, record: Record);
+	/// Indicates the start of a span of computation.
+	/// The returned object will be dropped when the span ends.
+	fn start(&self, span: Span, parent: Option<&Self::UserSpan>) -> Self::UserSpan;
```
I was thinking a tree of spans would allow us to fully account for the entire lifetime duration of an HTLC, given enough spans. It would indeed not contain real/unique channel monitor persistence spans due to the HTLC:channel relation, it would be only from the perspective of each HTLC. Would that miss part of the "why"?

The tree-based approach was chosen to be compatible with [tracing implementations](https://opentelemetry.io/docs/concepts/signals/traces/) and get out-of-the-box visualizations, but additional relationships could be represented as id attributes on the `Span` variants I think. To your point about the mapping, perhaps HTLC traces and channel monitor persistence traces can be created independently, and correlated based on monitor update id attributes.

> I'd naively expect a parent OutboundHTLC span on a WaitingOnMonitorPersist span because the outbound HTLC is waiting on a monitor persist for some part of its life

This is mostly the approach I took for the latest commits. Is the confusing part that each HTLC would have its own denormalized `WaitingOnMonitorPersist` span?

joostjager commented:
Linking https://github.com/lightningdevkit/rust-lightning/pull/4211 where an attempt is made to get rid of logging boiler plate and also enable the use of spans through standard `tracing`.
